### PR TITLE
state/metrics: Add concept of saving a model metric

### PR DESCRIPTION
**Pull request by mattyw**
## Description of change

In an upcoming pr we will add support for machines and models being able to record metrics. This pr adds the supporting functions in state

## QA steps

As this is an internal change only there is nothing to QA. That will occur in a follow up

## Documentation changes

The presence of model metrics might be something we wish to document. But how and why is not decided yet

## Bug reference

n/a

### CODE REVIEW

File: state/metrics.go [hunk 1]
```diff
@@ -161,6 +173,52 @@ func (st *State) AddMetrics(batch BatchParam) (*MetricBatch, error) {
...
+		doc: metricBatchDoc{
+			UUID:           batch.UUID,
+			ModelUUID:      st.ModelUUID(),
+			Sent:           false,
+			Created:        batch.Created,
+			Metrics:        batch.Metrics,
+			SLACredentials: slaCreds,
+		},
+	}
+	buildTxn := func(attempt int) ([]txn.Op, error) {
```
Comment (alesstimec) [inline state/metrics.go]: do we really have to use transactions to insert metrics into a collection? seems a bit of an overkill...

File: state/metrics.go [hunk 2]
```diff
@@ -34,15 +34,16 @@ type MetricBatch struct {
...
+	UUID           string    `bson:"_id"`
+	ModelUUID      string    `bson:"model-uuid"`
+	Unit           string    `bson:"unit"`
+	CharmURL       string    `bson:"charmurl"`
+	Sent           bool      `bson:"sent"`
+	DeleteTime     time.Time `bson:"delete-time"`
+	Created        time.Time `bson:"created"`
+	Metrics        []Metric  `bson:"metrics"`
+	Credentials    []byte    `bson:"credentials"`
+	SLACredentials []byte    `bson:"sla-Credentials"`
```
Comment (alesstimec) [inline state/metrics.go]: `bson:"sla-credentials,omitempty"`     if not, then at least lowercase the Credentials in the bson tag

File: state/metrics.go (hunk 1 above)
Comment (mattyw) [inline state/metrics.go]: @alesstimec as discussed I'm afraid we do

**Comment by mattyw** [comment]
$$merge$$

**Pull request merged**